presto-hive/src/main/java/com/facebook/presto/hive/IonSqlQueryBuilder.java

```diff
@@ -0,0 +1,175 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.StandardErrorCode;
+import com.facebook.presto.spi.predicate.Domain;
+import com.facebook.presto.spi.predicate.Range;
+import com.facebook.presto.spi.predicate.TupleDomain;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.util.stream.Collectors.joining;
+
+/**
+ * S3 Select uses Ion SQL++ query language. This class is used to construct a valid Ion SQL++ query
+ * to be evaluated with S3 Select on an S3 object.
+ */
+public class IonSqlQueryBuilder
+        extends QueryBuilder
+{
+    public IonSqlQueryBuilder(TypeManager typeManager)
+    {
+        super(typeManager);
+    }
+
+    public String buildSql(List<HiveColumnHandle> columns, TupleDomain<HiveColumnHandle> tupleDomain)
+    {
+        StringBuilder sql = new StringBuilder();
+
+        sql.append("SELECT ");
+
+        if (columns.isEmpty()) {
+            sql.append("' '");
+        }
+        else {
+            String columnNames = columns.stream()
+                    .map(column -> String.format("s._%d", column.getHiveColumnIndex() + 1))
+                    .collect(joining(", "));
+            sql.append(columnNames);
+        }
+
+        sql.append(" FROM ");
+        sql.append(DATA_SOURCE);
+
+        List<String> clauses = toConjuncts(columns, tupleDomain);
+        if (!clauses.isEmpty()) {
+            sql.append(" WHERE ")
+                    .append(Joiner.on(" AND ").join(clauses));
+        }
+
+        return sql.toString();
+    }
+
+    private List<String> toConjuncts(List<HiveColumnHandle> columns, TupleDomain<HiveColumnHandle> tupleDomain)
+    {
+        ImmutableList.Builder<String> builder = ImmutableList.builder();
+        for (HiveColumnHandle column : columns) {
+            Type type = column.getHiveType().getType(typeManager);
+            if (isAcceptedType(type)) {
+                Domain domain = tupleDomain.getDomains().get().get(column);
+                if (domain != null && shouldPushDown(domain)) {
+                    builder.add(toPredicate(domain, type, column.getHiveColumnIndex()));
+                }
+            }
+            else {
+                throw new PrestoException(StandardErrorCode.NOT_SUPPORTED, "unsupported type " + type.getDisplayName() + " for column " + column.getName());
```

Comment:
This should simply result in "unenforced predicate" rather than exception. Then you can probably remove timestamp from `isAcceptedType` and remove `shouldPushDown()` method (be sure to preserve the comment you placed there, it's informative)

Reply:
Makes sense, we were experimenting with that code block to skip column types during testing without affecting other parts of the code. We don't need it for PR, refactored.

## Test 

### Schema
```
create external table test_timestamp
(
 c1 timestamp,
 c2 decimal,
 c3 double
) ROW format delimited fields terminated BY ','
LOCATION 's3://sameerch-presto-test/test-timestamp';
```

### Input: s3://sameerch-presto-test/test-timestamp
```
2015-04-23 21:08:34.152923000,10,1.0
2015-04-24 21:08:34.152923000,9,6.7
```

### Query with timestamp in filter predicate
```
select * from test_timestamp where c1 > timestamp '2015-04-20';
```

### Expected Pushdown query 
```
select s._1, s._2, s._3 from S3Object s;
```

### Actual Pushdown query 
```
select s._1, s._2, s._3 from S3Object s;
```